### ISSUE TITLE
Unicode character escape sequences in interpolated string literals

### ISSUE THREAD

**Issue by RexJaeschke** [dotnet/csharpstandard#41]
For V6, I spec'd interpolated string literals using the text and grammar from the MS V6 spec. Today, while researching "all-things-Unicode sequences" in the ANTLR grammar I came up with the following questions:

Q1: In 7.4.2, "Unicode character escape sequences", we say

> A Unicode escape sequence represents a Unicode code point. Unicode escape sequences are processed in identifiers (§7.4.3), character literals (§7.4.5.5), and regular string literals (§7.4.5.6). A Unicode escape sequence is ***not*** processed in any other location (for example, to form an operator, punctuator, or keyword).

Notably absent from the list of allowed places is interpolated regular string literals (7.4.5.7). I see no reason for them to be omitted; besides, the MS V6 grammar I copied *does* allow *unicode_escape_sequence* in *interpolated_regular_string_character*. And my experiments with the latest version of the VS compiler show they are supported.

So, unless someone thinks otherwise, I'll add the missing item to the text in 7.4.2.

Q2: An interpolated string literal has the general form `$"  { xxx } "`, and we can escape a brace by doubling it (`{{`/`}}`). The grammar has these brace characters hardcoded as `'{'` and `'}'`, which means that Unicode sequences are *not* recognized there. And the VS compiler confirms that. Should they be allowed there?

FYI, here are the tests I ran, with the results ('\u007B' is left-brace, '\u007D' is right-brace):

```csharp
int x = 100;
Console.WriteLine($">{x}<");        // OK; writes out ">100<"
Console.WriteLine($"\u003E{\u0078}\u003C");   // OK, Unicode sequence for '>', identifier 'x', and '<'
Console.WriteLine($">\u007Bx}<");   // Rejected; CS8087: A '{' character may only be escaped by doubling '{{' in an interpolated string.
Console.WriteLine($">{x\u007D<");   // Rejected; CS8076: Missing close delimiter '}' for interpolated expression started with '{'

Console.WriteLine($">{{x}}");   // OK; writes out ">x<"
Console.WriteLine($">\u007B{x}}");   // Rejected; CS8087:
Console.WriteLine($">{\u007Bx}}"); // Rejected; CS1525: Invalid expression term "
Console.WriteLine($">{{x\u007D}");   // Rejected; CS8087:
Console.WriteLine($">{{x}\u007D");   // Rejected; CS8086: 
Console.WriteLine($">{{x\u007D\u007D");   // Rejected; CS8087:
```
Q3: Can Unicode sequences go between a single `{` and `}`? Yes, they can be used in identifiers, character and string literals. However, the in-between text need not be an expression; for example:

```csharp
Console.WriteLine($"{Math.PI , FieldWidthRightAligned : F3} - display only three decimal digits of the pi number");
```
Could I use a Unicode escape for the comma or colon? The VS compiler says no! What do you think?

**Comment by RexJaeschke** [comment]
@jskeet I think your comment argues against doing anything re Q2 and Q3, which is fine with me. But I think Q1 is still unaddressed.

**Comment by jskeet** [comment]
@RexJaeschke: Yes, it does make sense to include it in interpolated string literals, definitely.

**Comment by RexJaeschke** [comment]
Q1. was resolved by PR #103. Q2. and Q3. were answered *without* any changes needed.
